src/gui/MainFrame.cpp

```diff
@@ -557,8 +557,62 @@ void MyMainFrame::OpenProject(wxString project_filename) {
                 my_dialog->ShowDetailedText(changes);
             }
             if ( my_dialog->ShowModal( ) == wxID_YES ) {
-                my_dialog->Destroy( );
-                current_project.database.UpdateSchema(schema_comparison.second);
+
+                // 1. Estimate number of rows that will be updated
+                // Total number of particles can be very large; on the order of billions.
+                // Use long to ensure enough space.
+                unsigned long row_updates = 0;
+                int           normalized_increments; // This will adjust how long it takes for an update based on the total number of particles that will be processed
+
+                // Limit scope of temporarily needed arrays while estimating total number of increments...
+                {
+                    wxArrayInt ref_pkg_ids = current_project.database.ReturnIntArrayFromSelectCommand("select REFINEMENT_PACKAGE_ASSET_ID from REFINEMENT_PACKAGE_ASSETS");
+                    for ( int cur_pkg_id : ref_pkg_ids ) {
+                        row_updates += current_project.database.ReturnSingleIntFromSelectCommand(wxString::Format("select COUNT(*) from REFINEMENT_PACKAGE_CONTAINED_PARTICLES_%i", cur_pkg_id));
+                    }
+                    wxArrayString refinement_result_tables = current_project.database.ReturnStringArrayFromSelectCommand("SELECT name FROM sqlite_master WHERE type='table' AND name LIKE 'REFINEMENT_RESULT_%_%'");
+                    for ( wxString table_name : refinement_result_tables ) {
+                        row_updates += current_project.database.ReturnSingleIntFromSelectCommand(wxString::Format("SELECT COUNT(*) FROM '%s'", table_name));
+                    }
+                    wxArrayString classification_result_tables = current_project.database.ReturnStringArrayFromSelectCommand("SELECT name FROM sqlite_master WHERE type='table' AND name LIKE 'CLASSIFICATION_RESULT_%'");
+                    for ( wxString table_name : classification_result_tables ) {
+                        row_updates += current_project.database.ReturnSingleIntFromSelectCommand(wxString::Format("SELECT COUNT(*) FROM '%s'", table_name));
+                    }
+                } // End total increments estimation
+
+                row_updates += schema_comparison.second.size( );
+
+                // Adjusting the normalized increments based on total increments helps prevent GUI from freezing up and gives more accurate
+                // estimated time remaining. This may need additional fine tuning.
+                {
+                    long tmp_row_updates = row_updates;
```

Comment:
You are using a long here, and setting it from an unsigned long. It would probably be better to pick one and stick to it. Is there any chance that row updates would exhaust a long? If I'd err toward signed over un-signed (and certainly not a mix.)

Reply:
Ah, this was a mistake; when working on this a while ago, I likely used unsigned since the value should never be negative, then later decided signed long was plenty big enough. Will fix this.